"use strict";(globalThis.webpackChunkmy_website=globalThis.webpackChunkmy_website||[]).push([[4510],{3222(e,i,t){t.r(i),t.d(i,{assets:()=>l,contentTitle:()=>o,default:()=>h,frontMatter:()=>a,metadata:()=>n,toc:()=>c});const n=JSON.parse('{"id":"analysis/tool_lists/External studies/CircNetVis","title":"CircNetVis","description":"Abstract","source":"@site/docs/4.analysis/tool_lists/2.External studies/1.CircNetVis.md","sourceDirName":"4.analysis/tool_lists/2.External studies","slug":"/analysis/tool_lists/External studies/CircNetVis","permalink":"/river-docs/docs/analysis/tool_lists/External studies/CircNetVis","draft":false,"unlisted":false,"tags":[],"version":"current","sidebarPosition":1,"frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"Overview","permalink":"/river-docs/docs/analysis/tool_lists/External studies/Overview"},"next":{"title":"TF-Finder","permalink":"/river-docs/docs/analysis/tool_lists/External studies/TF-finder"}}');var s=t(4848),r=t(8453);const a={},o=void 0,l={},c=[{value:"Abstract",id:"abstract",level:2},{value:"Citation",id:"citation",level:2},{value:"RIVER integration",id:"river-integration",level:2},{value:"Step 1: Fork the Repository",id:"step-1-fork-the-repository",level:3},{value:"Step 2: Clone the Repository",id:"step-2-clone-the-repository",level:3},{value:"Step 3: Add Testing Infrastructure",id:"step-3-add-testing-infrastructure",level:3},{value:"Step 4: Create the RIVER Integration",id:"step-4-create-the-river-integration",level:3},{value:"Step 5: Test Locally",id:"step-5-test-locally",level:3},{value:"Step 6: Release on the platform",id:"step-6-release-on-the-platform",level:3}];function d(e){const i={a:"a",admonition:"admonition",code:"code",h2:"h2",h3:"h3",img:"img",li:"li",p:"p",pre:"pre",strong:"strong",ul:"ul",...(0,r.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsx)(i.h2,{id:"abstract",children:"Abstract"}),"\n",(0,s.jsxs)(i.p,{children:["Analyzing the interactions of circular RNAs (circRNAs) is a crucial step in understanding their functional impacts. While there are numerous visualization tools available for investigating circRNA interaction networks, these tools are typically limited to known circRNAs from specific databases. Moreover, these existing tools usually require complex installation procedures which can be time-consuming and challenging for users. There is a lack of a user-friendly web application that facilitates interactive exploration and visualization of circRNA interaction networks. CircNetVis is an interactive online web application to enhance the analysis of human/mouse circRNA interactions. The tool allows three different input formats of circRNAs including circRNA IDs from CircBase, circRNA coordinates (chromosome, start position, end position), and circRNA sequences in the FASTA format. It integrates multiple interaction networks for visualization and investigation of the interplay between circRNA, microRNAs, mRNAs and RNA binding proteins. CircNetVis also enables users to interactively explore the interactions of unknown circRNAs which are not reported from previous databases. The tool can generate interactive plots and allows users to save results as output files for offline usage. CircNetVis is implemented as a web application using R-shiny and freely available for academic use at ",(0,s.jsx)(i.a,{href:"https://www.meb.ki.se/shiny/truvu/CircNetVis/",children:"https://www.meb.ki.se/shiny/truvu/CircNetVis/"}),"."]}),"\n",(0,s.jsx)(i.h2,{id:"citation",children:"Citation"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bibtex",children:"@article{nguyen2024circnetvis,\n  title={CircNetVis: an interactive web application for visualizing interaction networks of circular RNAs},\n  author={Nguyen, Thi-Hau and Nguyen, Ha-Nam and Vu, Trung Nghia},\n  journal={BMC bioinformatics},\n  volume={25},\n  number={1},\n  pages={31},\n  year={2024},\n  publisher={Springer}\n}\n"})}),"\n",(0,s.jsx)(i.h2,{id:"river-integration",children:"RIVER integration"}),"\n",(0,s.jsx)(i.admonition,{type:"info",children:(0,s.jsxs)(i.ul,{children:["\n",(0,s.jsx)(i.li,{children:"It is the public tools, new configuration is added to run on RIVER"}),"\n",(0,s.jsxs)(i.li,{children:["Availble at: ",(0,s.jsx)(i.a,{href:"https://github.com/riverxdata/bioinfor-sw-CircNetVis",children:"https://github.com/riverxdata/bioinfor-sw-CircNetVis"})]}),"\n"]})}),"\n",(0,s.jsx)(i.p,{children:(0,s.jsx)(i.img,{alt:"CirNetVis",src:t(6153).A+"",width:"2560",height:"1413"})}),"\n",(0,s.jsx)(i.h3,{id:"step-1-fork-the-repository",children:"Step 1: Fork the Repository"}),"\n",(0,s.jsxs)(i.p,{children:["Fork TFFinder if you're not the owner. See ",(0,s.jsx)(i.a,{href:"https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo",children:"GitHub forking documentation"}),"."]}),"\n",(0,s.jsx)(i.h3,{id:"step-2-clone-the-repository",children:"Step 2: Clone the Repository"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"git clone git@github.com:riverxdata/bioinfor-sw-CircNetVis.git\ncd bioinfor-sw-CircNetVis\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-3-add-testing-infrastructure",children:"Step 3: Add Testing Infrastructure"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"git clone git@github.com:riverxdata/web-template-analysis.git\ncp -r web-template-analysis/tests tests\nrm -rf web-template-analysis\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-4-create-the-river-integration",children:"Step 4: Create the RIVER Integration"}),"\n",(0,s.jsxs)(i.p,{children:["Create ",(0,s.jsx)(i.code,{children:"river/main.sh"})," with the following content:"]}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:'cd analysis\n# Download the database files\nif [ -d "CircNetVis/Tools" ]; then\n    echo "Database already exists."\nelse\n    wget https://zenodo.org/records/10394426/files/CircNetVis_database.tar.gz?download=1 -O CircNetVis_database.tar.gz\n    tar -xvzf CircNetVis_database.tar.gz -C CircNetVis --strip-components 1\nfi\n# Run the Shiny app\neval "$(pixi shell-hook)"\nRscript -e "shiny::runApp(\'CircNetVis_mm9\')"\n'})}),"\n",(0,s.jsxs)(i.p,{children:[(0,s.jsx)(i.strong,{children:"Final directory structure:"}),"\nFor the sub folder ",(0,s.jsx)(i.code,{children:"river"}),", it is similar to previous python web application"]}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"\u251c\u2500\u2500 CircNetVis\n\u251c\u2500\u2500 CircNetVis_mm9\n\u251c\u2500\u2500 License\n\u251c\u2500\u2500 pixi.lock\n\u251c\u2500\u2500 pixi.toml\n\u251c\u2500\u2500 README.html\n\u251c\u2500\u2500 README.md\n\u251c\u2500\u2500 river\n\u251c\u2500\u2500 tests\n\u2514\u2500\u2500 work\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-5-test-locally",children:"Step 5: Test Locally"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"bash tests/run_local.sh\n"})}),"\n",(0,s.jsx)(i.p,{children:"The terminal will display access URLs:"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"CircNetVis_mm9 resources already exist.\nLoading required package: shiny\n\nListening on http://127.0.0.1:3411\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-6-release-on-the-platform",children:"Step 6: Release on the platform"}),"\n",(0,s.jsx)(i.p,{children:"There are 2 models: Human vs mouse\nInstead of creating the paramter to select which Shiny app is used, it can be simply tagged with different model\nOn the RIVER UI, you can select the tag to deploy your specific model. After commit the repository, you can create and push your tag to github"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"git tag v1.0.0\ngit push origin main\ngit push origin v1.0.0\n"})})]})}function h(e={}){const{wrapper:i}={...(0,r.R)(),...e.components};return i?(0,s.jsx)(i,{...e,children:(0,s.jsx)(d,{...e})}):d(e)}},6153(e,i,t){t.d(i,{A:()=>n});const n=t.p+"assets/images/circvisnet-84f4b63ade229531a5684507814b3bba.png"},8453(e,i,t){t.d(i,{R:()=>a,x:()=>o});var n=t(6540);const s={},r=n.createContext(s);function a(e){const i=n.useContext(r);return n.useMemo(function(){return"function"==typeof e?e(i):{...i,...e}},[i,e])}function o(e){let i;return i=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:a(e.components),n.createElement(r.Provider,{value:i},e.children)}}}]);