"use strict";(globalThis.webpackChunkmy_website=globalThis.webpackChunkmy_website||[]).push([[2878],{1061(e,i,t){t.r(i),t.d(i,{assets:()=>l,contentTitle:()=>o,default:()=>h,frontMatter:()=>a,metadata:()=>n,toc:()=>d});const n=JSON.parse('{"id":"analysis/tool_lists/External studies/TF-finder","title":"TF-Finder","description":"Abstract","source":"@site/docs/4.analysis/tool_lists/2.External studies/2.TF-finder.md","sourceDirName":"4.analysis/tool_lists/2.External studies","slug":"/analysis/tool_lists/External studies/TF-finder","permalink":"/river-docs/docs/analysis/tool_lists/External studies/TF-finder","draft":false,"unlisted":false,"tags":[],"version":"current","sidebarPosition":2,"frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"CircNetVis","permalink":"/river-docs/docs/analysis/tool_lists/External studies/CircNetVis"},"next":{"title":"Overview","permalink":"/river-docs/docs/analysis/tool_lists/nf-core/Overview"}}');var s=t(4848),r=t(8453);const a={},o="TF-Finder",l={},d=[{value:"Abstract",id:"abstract",level:2},{value:"Citation",id:"citation",level:2},{value:"RIVER integration",id:"river-integration",level:2},{value:"Step 1: Fork the Repository",id:"step-1-fork-the-repository",level:3},{value:"Step 2: Clone the Repository",id:"step-2-clone-the-repository",level:3},{value:"Step 3: Add Testing Infrastructure",id:"step-3-add-testing-infrastructure",level:3},{value:"Step 4: Create the RIVER Integration",id:"step-4-create-the-river-integration",level:3},{value:"Step 5: Test Locally",id:"step-5-test-locally",level:3},{value:"Step 6: Release",id:"step-6-release",level:3}];function c(e){const i={a:"a",admonition:"admonition",code:"code",h1:"h1",h2:"h2",h3:"h3",header:"header",img:"img",li:"li",p:"p",pre:"pre",ul:"ul",...(0,r.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsx)(i.header,{children:(0,s.jsx)(i.h1,{id:"tf-finder",children:"TF-Finder"})}),"\n",(0,s.jsx)(i.h2,{id:"abstract",children:"Abstract"}),"\n",(0,s.jsxs)(i.p,{children:["Transcription is a key cell process that consists of synthesizing several copies of RNA from a gene DNA sequence. This process is highly regulated and closely linked to the ability of transcription factors to bind specifically to DNA. TFinder is an easy-to-use Python web portal allowing the identification of Individual Motifs (IM) such as Transcription Factor Binding Sites (TFBS).\nUsing the NCBI API, TFinder extracts either promoter or gene terminal regulatory regions, through a simple query of NCBI gene name or ID. It enables simultaneous analysis across five different species for an unlimited number of genes. TFinder searches for Individual Motifs in different formats, including IUPAC codes and JASPAR entries. Moreover, TFinder also allows de novo generations of a Position Weight Matrix (PWM) and the use of already established PWM. Finally, the data are provided in a tabular and a graph format showing the relevance and the P-value of the Individual Motifs found as well as their location relative to the Transcription Start Site (TSS) or the terminal region of the gene. The results are then sent by email to users facilitating the subsequent data analysis and sharing.\nTFinder is written in Python and freely available on GitHub under the MIT license: ",(0,s.jsx)(i.a,{href:"https://github.com/Jumitti/TFinder",children:"https://github.com/Jumitti/TFinder"}),". It can be accessed as a web application implemented in Streamlit at ",(0,s.jsx)(i.a,{href:"https://tfinder-ipmc.streamlit.app",children:"https://tfinder-ipmc.streamlit.app"}),". Resources are available on Streamlit \u201cResources\u201d tab. TFINDER strength is that it relies on an all-in-one intuitive tool allowing users inexperienced with bioinformatics tools to retrieve gene regulatory regions sequences in multiple species and to search for individual motifs in a huge number of genes."]}),"\n",(0,s.jsx)(i.h2,{id:"citation",children:"Citation"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bibtex",children:"@article{minniti2025tfinder,\n  title={TFinder: A Python Web Tool for Predicting Transcription Factor Binding Sites},\n  author={Minniti, Julien and Checler, Frederic and Duplan, Eric and da Costa, Cristine Alves},\n  journal={Journal of Molecular Biology},\n  volume={437},\n  number={3},\n  pages={168921},\n  year={2025},\n  publisher={Elsevier}\n}\n"})}),"\n",(0,s.jsx)(i.h2,{id:"river-integration",children:"RIVER integration"}),"\n",(0,s.jsx)(i.admonition,{type:"info",children:(0,s.jsxs)(i.ul,{children:["\n",(0,s.jsx)(i.li,{children:"It is the public tools, new configuration is added to run on RIVER"}),"\n",(0,s.jsxs)(i.li,{children:["Available at: ",(0,s.jsx)(i.a,{href:"https://github.com/riverxdata/bioinfor-sw-TFinder/",children:"https://github.com/riverxdata/bioinfor-sw-TFinder/"})]}),"\n"]})}),"\n",(0,s.jsx)(i.p,{children:(0,s.jsx)(i.img,{alt:"CirNetVis",src:t(9868).A+"",width:"2560",height:"1413"})}),"\n",(0,s.jsxs)(i.p,{children:["For existing bioinformatics tools with web interfaces, simply add a ",(0,s.jsx)(i.code,{children:"river/main.sh"})," file. Recommended approach: use pixi for environment management or singularity for containerized tools."]}),"\n",(0,s.jsx)(i.p,{children:"TFFinder is a Streamlit application that demonstrates the adaptation process."}),"\n",(0,s.jsx)(i.h3,{id:"step-1-fork-the-repository",children:"Step 1: Fork the Repository"}),"\n",(0,s.jsxs)(i.p,{children:["Fork TFFinder if you're not the owner. See ",(0,s.jsx)(i.a,{href:"https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo",children:"GitHub forking documentation"}),"."]}),"\n",(0,s.jsx)(i.h3,{id:"step-2-clone-the-repository",children:"Step 2: Clone the Repository"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"git clone git@github.com:riverxdata/bioinfor-sw-TFinder.git\ncd bioinfor-sw-TFinder\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-3-add-testing-infrastructure",children:"Step 3: Add Testing Infrastructure"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"git clone git@github.com:riverxdata/web-template-analysis.git\ncp -r web-template-analysis/tests tests\nrm -rf web-template-analysis\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-4-create-the-river-integration",children:"Step 4: Create the RIVER Integration"}),"\n",(0,s.jsxs)(i.p,{children:["Create ",(0,s.jsx)(i.code,{children:"river/main.sh"})," with the following content:"]}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:'cd $RIVER_HOME/jobs/$job_id/analysis\nENV=$repo_name-$tag\neval "$(pixi shell-hook --manifest-path analysis)"\nstreamlit run TFinder-v1.py --server.port $PORT --server.headless true\n'})}),"\n",(0,s.jsx)(i.p,{children:"Final directory structure:"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{children:"\u2500\u2500 clock.time\n\u251c\u2500\u2500 img\n\u251c\u2500\u2500 LICENSE\n\u251c\u2500\u2500 navigation\n\u251c\u2500\u2500 README.md\n\u251c\u2500\u2500 requirements.txt\n\u251c\u2500\u2500 SECURITY.md\n\u251c\u2500\u2500 streamlit_modal\n\u2502   \u2514\u2500\u2500 __init__.py\n\u251c\u2500\u2500 tests\n\u2502   \u251c\u2500\u2500 params.json\n\u2502   \u2514\u2500\u2500 run_local.sh\n\u251c\u2500\u2500 tfinder\n\u2502   \u2514\u2500\u2500 __init__.py\n\u251c\u2500\u2500 TFinder_health_universe.py\n\u251c\u2500\u2500 TFinder_standalone_win.bat\n\u251c\u2500\u2500 TFinder-v1.py\n\u251c\u2500\u2500 utils\n\u2502   \u2514\u2500\u2500 components.py\n\u251c\u2500\u2500 river\n\u2502   \u2514\u2500\u2500 main.sh\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-5-test-locally",children:"Step 5: Test Locally"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"bash tests/run_local.sh\n"})}),"\n",(0,s.jsx)(i.p,{children:"The terminal will display access URLs:"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{children:"Local URL: http://localhost:39905\nNetwork URL: http://192.168.1.5:39905\nExternal URL: http://1.54.25.47:39905\n"})}),"\n",(0,s.jsx)(i.h3,{id:"step-6-release",children:"Step 6: Release"}),"\n",(0,s.jsx)(i.p,{children:"It has to add tag so the RIVER platform can fetch with specific version"}),"\n",(0,s.jsx)(i.pre,{children:(0,s.jsx)(i.code,{className:"language-bash",children:"git tag v1.0.0\ngit push origin main\ngit push origin v1.0.0\n"})})]})}function h(e={}){const{wrapper:i}={...(0,r.R)(),...e.components};return i?(0,s.jsx)(i,{...e,children:(0,s.jsx)(c,{...e})}):c(e)}},8453(e,i,t){t.d(i,{R:()=>a,x:()=>o});var n=t(6540);const s={},r=n.createContext(s);function a(e){const i=n.useContext(r);return n.useMemo(function(){return"function"==typeof e?e(i):{...i,...e}},[i,e])}function o(e){let i;return i=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:a(e.components),n.createElement(r.Provider,{value:i},e.children)}},9868(e,i,t){t.d(i,{A:()=>n});const n=t.p+"assets/images/tf-finder-1bffac0ab4cc29238a0d04b9760f4efc.png"}}]);