"use strict";(globalThis.webpackChunkmy_website=globalThis.webpackChunkmy_website||[]).push([[970],{798(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/PR_GHA-3f6f906f7d1600d69ba6fe6bd331ffe3.png"},1341(e,n,i){i.r(n),i.d(n,{assets:()=>l,contentTitle:()=>r,default:()=>h,frontMatter:()=>o,metadata:()=>t,toc:()=>c});var t=i(9264),s=i(4848),a=i(8453);const o={slug:"how-to-version-control-git-bioinformatics-part-2",title:"The Evolution of Version Control - Git's Role in Reproducible Bioinformatics (Part 2)",authors:["river"],tags:["git","version-control","bioinformatics","reproducibility","pipelines","ci-cd","github-actions"],image:"./imgs/github_intro.svg"},r=void 0,l={image:i(4051).A,authorsImageUrls:[void 0]},c=[{value:"Git commands",id:"git-commands",level:2},{value:"How to show commit log",id:"how-to-show-commit-log",level:3},{value:"How to checkout to specific commit",id:"how-to-checkout-to-specific-commit",level:3},{value:"How to view the difference between two commits or your current state",id:"how-to-view-the-difference-between-two-commits-or-your-current-state",level:3},{value:"How to roll back",id:"how-to-roll-back",level:3},{value:"Add data",id:"add-data",level:2},{value:"Add workflow",id:"add-workflow",level:2},{value:"Add workflow main.sh script and testing",id:"add-workflow-mainsh-script-and-testing",level:3},{value:"Use Makefile to add tests",id:"use-makefile-to-add-tests",level:3},{value:"Configure testing",id:"configure-testing",level:2},{value:"Create GitHub Action files",id:"create-github-action-files",level:3},{value:"Running GitHub Actions locally",id:"running-github-actions-locally",level:3},{value:"What is CI/CD in bioinformatics?",id:"what-is-cicd-in-bioinformatics",level:2},{value:"What is CI/CD in software development",id:"what-is-cicd-in-software-development",level:3},{value:"How does it work in bioinformatics?",id:"how-does-it-work-in-bioinformatics",level:3},{value:"Review",id:"review",level:2},{value:"Git Commands for Project Management",id:"git-commands-for-project-management",level:3},{value:"Makefile Automation",id:"makefile-automation",level:3},{value:"Workflow Scripts",id:"workflow-scripts",level:3},{value:"CI/CD with GitHub Actions",id:"cicd-with-github-actions",level:3},{value:"Local Testing with <code>act</code>",id:"local-testing-with-act",level:3},{value:"CI/CD in Bioinformatics Context",id:"cicd-in-bioinformatics-context",level:3}];function d(e){const n={a:"a",admonition:"admonition",code:"code",h2:"h2",h3:"h3",img:"img",li:"li",p:"p",pre:"pre",strong:"strong",ul:"ul",...(0,a.R)(),...e.components};return(0,s.jsxs)(s.Fragment,{children:[(0,s.jsxs)(n.p,{children:["Welcome to Part 2 of our series on version control in bioinformatics. In ",(0,s.jsx)(n.a,{href:"/river-docs/blog/how-to-version-control-git-bioinformatics-part-1",children:"Part 1"}),", we introduced Git fundamentals, branching strategies, and collaborative workflows. In this post, we'll dive into how Continuous Integration and Continuous Deployment (CI/CD) can transform your bioinformatics projects. If these concepts are new to you, don't worry\u2014this guide will walk you through managing your bioinformatics repository to ensure your work is easily reproducible on any machine. Whether your server is wiped or you need to spin up a new virtual machine, you'll be able to quickly rerun your pipeline. With CI/CD, every code update can automatically trigger tests on a small dataset to verify everything works before scaling up, ensuring that new changes don't break your results or workflows."]}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Part 1"}),": History, basics, GitHub integration, and hands-on tutorials"]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Part 2 (This Post)"}),": Local testing, CI/CD implementation, automated testing, and reproducible pipelines"]}),"\n"]}),"\n",(0,s.jsxs)(n.admonition,{type:"info",children:[(0,s.jsx)(n.p,{children:"In this tutorial, we will:"}),(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Add code to download data"}),"\n",(0,s.jsx)(n.li,{children:"Write scripts to use FastQC and MultiQC"}),"\n"]})]}),"\n",(0,s.jsx)(n.h2,{id:"git-commands",children:"Git commands"}),"\n",(0,s.jsx)(n.p,{children:"This tutorial is fully committed via a GitHub repository. You can clone and checkout to specific branches to see how it is developed and committed.\nFirst, clone the repository, then:"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Learn how to view commit history (branches are referenced by their names)"}),"\n",(0,s.jsx)(n.li,{children:"Learn how to checkout to a specific commit"}),"\n",(0,s.jsx)(n.li,{children:"Learn how to see the difference between the current commit and a specific commit"}),"\n"]}),"\n",(0,s.jsx)(n.admonition,{type:"tip",children:(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["To quit the git prompt, type ",(0,s.jsx)(n.code,{children:"q"})]}),"\n",(0,s.jsx)(n.li,{children:"Create your own repository for a more practical approach"}),"\n",(0,s.jsx)(n.li,{children:"Check the differences in files, then copy them to your own repository to follow the tutorial"}),"\n"]})}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"# clone\ngit clone git@github.com:nttg8100/ngs-quality-control-for-blog-example.git\ncd ngs-quality-control-for-blog-example\n\n# checkout to branch for the relevant tutorial step, e.g. feature/add-data\ngit checkout feature/add-data\n"})}),"\n",(0,s.jsx)(n.h3,{id:"how-to-show-commit-log",children:"How to show commit log"}),"\n",(0,s.jsx)(n.p,{children:"Git provides commands to view the history and check the difference between two commits."}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git log\n"})}),"\n",(0,s.jsx)(n.p,{children:"It will show details as below:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"commit ca29351295a4148bda73b6bf76a42cc7b834d1b6 (HEAD -> feature/add-data, origin/feature/add-data)\nAuthor: Thanh-Giang (River) Tan Nguyen <nttg8100@gmail.com>\nDate:   Mon Jan 19 23:34:58 2026 +0700\n\n    feat: add makefile to get data\n\ncommit 70d584a726fb5418274448a4f51ca07d94b104be (origin/main, origin/HEAD, main, feature/testing-rebase)\nMerge: c731b9a 230795b\nAuthor: Thanh Giang Nguyen Tan <64969412+nttg8100@users.noreply.github.com>\nDate:   Sat Jan 17 16:42:11 2026 +0700\n\n    Merge pull request #1 from nttg8100/feature/environment-setup\n    \n    feat: add pixi setup\n\ncommit 230795b7b8b03087ef7ee938d333161c9fcc0731 (origin/feature/environment-setup)\nAuthor: Thanh-Giang (River) Tan Nguyen <nttg8100@gmail.com>\nDate:   Sat Jan 17 15:05:28 2026 +0700\n\n    feat: add pixi setup\n\ncommit c731b9a80dfb22936bfcf94442bdf9b488a90d3c (feature/tesing-rebase)\nAuthor: Thanh Giang Nguyen Tan <64969412+nttg8100@users.noreply.github.com>\nDate:   Wed Jan 14 23:09:30 2026 +0700\n"})}),"\n",(0,s.jsx)(n.p,{children:"To see more precisely what you need:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git log --oneline\n"})}),"\n",(0,s.jsx)(n.p,{children:"The one-line format is good enough:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"ca29351 (HEAD -> feature/add-data, origin/feature/add-data) feat: add makefile to get data\n70d584a (origin/main, origin/HEAD, main, feature/testing-rebase) Merge pull request #1 from nttg8100/feature/environment-setup\n230795b (origin/feature/environment-setup) feat: add pixi setup\nc731b9a (feature/tesing-rebase) Initial commit\n"})}),"\n",(0,s.jsx)(n.p,{children:"Now you can see what I did on the repository. Following this tutorial, you can change to a specific commit to track it."}),"\n",(0,s.jsx)(n.h3,{id:"how-to-checkout-to-specific-commit",children:"How to checkout to specific commit"}),"\n",(0,s.jsxs)(n.p,{children:["You can simply use the commit hash. For example, if I want to see all files from commit ",(0,s.jsx)(n.code,{children:"230795b"})," which is on the branch ",(0,s.jsx)(n.code,{children:"feature/testing-rebase"}),":"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout 230795b\n"})}),"\n",(0,s.jsx)(n.p,{children:"Now your code editor will show the files with content related to this commit."}),"\n",(0,s.jsx)(n.h3,{id:"how-to-view-the-difference-between-two-commits-or-your-current-state",children:"How to view the difference between two commits or your current state"}),"\n",(0,s.jsx)(n.p,{children:"If you have made changes to your files but have not committed them yet:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git diff\n"})}),"\n",(0,s.jsx)(n.p,{children:"I made a change and have not committed it here:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"diff --git a/Makefile b/Makefile\nindex ae08113..c2e5b87 100644\n--- a/Makefile\n+++ b/Makefile\n@@ -10,4 +10,6 @@ data/GSE110004:\n        wget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_2.fastq.gz\n \n clean:\n-       rm -rf data/GSE110004\n\\ No newline at end of file\n+       rm -rf data/GSE110004\n+\n+Change here\n\\ No newline at end of file\n"})}),"\n",(0,s.jsx)(n.p,{children:"If you want to compare between two branches/commits:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git diff ca29351..70d584a\n"})}),"\n",(0,s.jsx)(n.p,{children:"The changes between the above commits:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"diff --git a/.gitignore b/.gitignore\nindex cfca1e8..1c3ecb3 100644\n--- a/.gitignore\n+++ b/.gitignore\n@@ -19,5 +19,3 @@ bin-release/\n # pixi environments\n .pixi/*\n !.pixi/config.toml\n-data\n-\ndiff --git a/Makefile b/Makefile\nindex ae08113..bcaa6d7 100644\n--- a/Makefile\n+++ b/Makefile\n@@ -1,13 +1,3 @@\n ${HOME}/.pixi/bin/pixi:\n        mkdir -p ${HOME}/.pixi/bin\n-       curl -sSL https://pixi.dev/install.sh | sh\n-\n-\n-data/GSE110004:\n-       mkdir -p data/GSE110004\n-       cd data/GSE110004 && \\\n-       wget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_1.fastq.gz && \\\n-       wget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_2.fastq.gz\n-\n-clean:\n-       rm -rf data/GSE110004\n\\ No newline at end of file\n+       curl -sSL https://pixi.dev/install.sh | sh\n\\ No newline at end of file\n"})}),"\n",(0,s.jsx)(n.h3,{id:"how-to-roll-back",children:"How to roll back"}),"\n",(0,s.jsxs)(n.p,{children:["If you checkout a specific commit and add a few changes to test according to the tutorial, ",(0,s.jsx)(n.code,{children:"git"})," requires you to commit your changes.\nYou can simply roll back to the original version:"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout -- .\n"})}),"\n",(0,s.jsx)(n.p,{children:"Sometimes you want to get files from another branch. You can use:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout <branch name> <relative file path>\n# ex: git checkout main .gitignore\n"})}),"\n",(0,s.jsx)(n.h2,{id:"add-data",children:"Add data"}),"\n",(0,s.jsx)(n.p,{children:"We need to create a new branch from updated main"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout main\ngit pull origin main\ngit switch -c feature/add-data\n"})}),"\n",(0,s.jsx)(n.p,{children:"Now we'll add a few target setups to the Makefile to automate data download and cleanup tasks for your pipeline:"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:[(0,s.jsxs)(n.strong,{children:[(0,s.jsx)(n.code,{children:"data/GSE110004"})," target"]}),": Downloads example FASTQ files from a public test dataset into the ",(0,s.jsx)(n.code,{children:"data/GSE110004"})," directory. This ensures everyone working on the project uses the same input data, improving reproducibility."]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsxs)(n.strong,{children:[(0,s.jsx)(n.code,{children:"clean"})," target"]}),": Removes the downloaded data directory, making it easy to reset your environment or save disk space."]}),"\n"]}),"\n",(0,s.jsx)(n.p,{children:"This approach helps standardize data management and makes it easier for collaborators (and CI/CD systems) to set up and test the pipeline with minimal manual steps."}),"\n",(0,s.jsxs)(n.p,{children:["Update your ",(0,s.jsx)(n.code,{children:"Makefile"})," with the following:"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"${HOME}/.pixi/bin/pixi:\n        mkdir -p ${HOME}/.pixi/bin\n        curl -sSL https://pixi.dev/install.sh | sh\n\ndata:\n  mkdir -p data/GSE110004\n  wget -O data/GSE110004/SRR6351937_1.fastq.gz https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR635/007/SRR6351937/SRR6351937_1.fastq.gz\n  wget -O data/GSE110004/SRR6351937_2.fastq.gz https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR635/007/SRR6351937/SRR6351937_2.fastq.gz\n\nclean:\n  rm -rf data\n"})}),"\n",(0,s.jsx)(n.p,{children:"Following tutorial 1, you can commit, push, create a PR, and merge to main:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:'git commit -m "feat: add makefile to get data"\ngit push origin feature/add-data\n'})}),"\n",(0,s.jsx)(n.h2,{id:"add-workflow",children:"Add workflow"}),"\n",(0,s.jsx)(n.p,{children:"Create new branch"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout main\ngit pull\ngit switch -c feature/workflow\n"})}),"\n",(0,s.jsx)(n.h3,{id:"add-workflow-mainsh-script-and-testing",children:"Add workflow main.sh script and testing"}),"\n",(0,s.jsx)(n.p,{children:"Here I create a simple bash script that will run using command lines with inputs as arguments."}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:'# create workflow file using bash script, this runs for only paired reads of 1 sample\ncat << EOF >> main.sh\n#!/bin/bash\n\n# Usage: ./main.sh <read1.fastq.gz> <read2.fastq.gz> <output_folder>\n\nset -e\n\nif [ "\\$#" -ne 3 ]; then\n    echo "Usage: $0 <read1.fastq.gz> <read2.fastq.gz> <output_folder>"\n    exit 1\nfi\n\nREAD1="\\$1"\nREAD2="\\$2"\nOUTDIR="\\$3"\n\nmkdir -p "\\$OUTDIR"\n\necho "Running FastQC on \\$READ1 and \\$READ2..."\nfastqc "\\$READ1" "\\$READ2" -o "\\$OUTDIR"\n\necho "Running MultiQC in \\$OUTDIR..."\nmultiqc "\\$OUTDIR" -o "\\$OUTDIR"\n\necho "All done. Results are in \\$OUTDIR"\nEOF\n\n# chmod to be executed\nchmod +x main.sh\n\n# get help script\nbash main.sh\n# Usage: zsh <read1.fastq.gz> <read2.fastq.gz> <output_folder>\n\n# test it with your data\n# activate pixi \npixi shell\n# run interactively\nbash main.sh data/GSE110004/*_1.fastq.gz data/GSE110004/*_2.fastq.gz result\n\n# run with pixi directly\npixi run bash main.sh data/GSE110004/*_1.fastq.gz data/GSE110004/*_2.fastq.gz result\n'})}),"\n",(0,s.jsx)(n.p,{children:(0,s.jsx)(n.img,{alt:"qc_wf",src:i(1595).A+"",width:"1519",height:"1036"})}),"\n",(0,s.jsx)(n.h3,{id:"use-makefile-to-add-tests",children:"Use Makefile to add tests"}),"\n",(0,s.jsxs)(n.p,{children:["Now we can quickly run tests after updating the ",(0,s.jsx)(n.code,{children:"Makefile"}),":"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-Makefile",children:".PHONY: install-pixi test-e2e clean\n${HOME}/.pixi/bin/pixi:\n\tmkdir -p ${HOME}/.pixi/bin\n\tcurl -sSL https://pixi.dev/install.sh | sh\n\ndata/GSE110004:\n\tmkdir -p data/GSE110004\n\tcd data/GSE110004 && \\\n\twget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_1.fastq.gz && \\\n\twget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_2.fastq.gz\n\ntest-e2e: data/GSE110004\n\t${HOME}/.pixi/bin/pixi run bash main.sh data/GSE110004/*_1.fastq.gz data/GSE110004/*_2.fastq.gz result\n\nclean:\n\trm -rf data/GSE110004\n"})}),"\n",(0,s.jsx)(n.p,{children:"Quickly run the test:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"make test-e2e\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Update your ",(0,s.jsx)(n.code,{children:".gitignore"})," to avoid committing the results:"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"# Build and Release Folders\nbin-debug/\nbin-release/\n[Oo]bj/\n[Bb]in/\n\n# Other files and folders\n.settings/\n\n# Executables\n*.swf\n*.air\n*.ipa\n*.apk\n\n# Project files, i.e. `.project`, `.actionScriptProperties` and `.flexProperties`\n# should NOT be excluded as they contain compiler settings and other important\n# information for Eclipse / Flash Builder.\n# pixi environments\n.pixi/*\n!.pixi/config.toml\ndata\nresult\n"})}),"\n",(0,s.jsx)(n.p,{children:"Following tutorial 1, you can commit, push, create a PR, and merge to main:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:'git add .\ngit commit -m "feat: add workflow"\ngit push origin feature/workflow \n'})}),"\n",(0,s.jsx)(n.p,{children:"Then, in your repository root folder, run these commands to get the data:"}),"\n",(0,s.jsx)(n.admonition,{type:"tip",children:(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Currently, a few make target commands are simple and can be run directly, but in the future, it would be better to update them with more commands"}),"\n",(0,s.jsxs)(n.li,{children:["Ex: ",(0,s.jsx)(n.code,{children:"make clean"})," can be used to delete more files/folders"]}),"\n"]})}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout \n# download data\nmake data\n# test that data is downloaded\nls -lah  data/GSE110004\n# total 4.3M\n# drwxrwxr-x 2 giangnguyen giangnguyen 4.0K Jan 19 01:00 .\n# drwxrwxr-x 3 giangnguyen giangnguyen 4.0K Jan 19 01:00 ..\n# -rw-rw-r-- 1 giangnguyen giangnguyen 2.2M Jan 19 01:00 SRR6357070_1.fastq.gz\n# -rw-rw-r-- 1 giangnguyen giangnguyen 2.2M Jan 19 01:00 SRR6357070_2.fastq.gz\n\n# delete data folder\nmake clean\n"})}),"\n",(0,s.jsx)(n.h2,{id:"configure-testing",children:"Configure testing"}),"\n",(0,s.jsxs)(n.p,{children:["So far, everything has been done locally. However, when you add new features\u2014for example, adding ",(0,s.jsx)(n.code,{children:"trimmomatic"})," to trim the FastQC files\u2014you need to run the tests again to ensure they can be executed without errors. This is the simplest testing practice for developing bioinformatics projects."]}),"\n",(0,s.jsx)(n.p,{children:"GitHub provides GitHub Actions to help with this:"}),"\n",(0,s.jsx)(n.admonition,{type:"info",children:(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["For more detailed examples on GitHub Actions, follow the official documentation: ",(0,s.jsx)(n.a,{href:"https://docs.github.com/en/actions/get-started/quickstart",children:"https://docs.github.com/en/actions/get-started/quickstart"})]}),"\n",(0,s.jsx)(n.li,{children:"Here, we introduce how it works and how we can apply it simply"}),"\n"]})}),"\n",(0,s.jsx)(n.h3,{id:"create-github-action-files",children:"Create GitHub Action files"}),"\n",(0,s.jsx)(n.p,{children:"Create a new branch for CI/CD:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"git checkout main\ngit pull\ngit switch -c feature/cicd\n"})}),"\n",(0,s.jsx)(n.p,{children:"Configure the workflow:"}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"mkdir -p .github/workflows\n# create workflow file\ncat << EOF >> .github/workflows/test-e2e.yaml\nname: Testing e2e\n\non:\n  push:\n    branches: [ main ]\n  pull_request:\n    branches: [ main ]\n\njobs:\n  clone:\n    runs-on: ubuntu-latest\n    steps:\n      - name: Checkout repository\n        uses: actions/checkout@v4\n\n      - name: Running testing\n        run: |\n          make test-e2e\nEOF\n"})}),"\n",(0,s.jsx)(n.p,{children:"What it does:"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Runs workflow named Testing e2e"}),"\n",(0,s.jsx)(n.li,{children:"Triggers on pull requests to main"}),"\n",(0,s.jsx)(n.li,{children:"Triggers on push or merge into main"}),"\n",(0,s.jsx)(n.li,{children:"Starts a job on ubuntu-latest runner"}),"\n",(0,s.jsx)(n.li,{children:"Clones the repository using actions/checkout"}),"\n",(0,s.jsx)(n.li,{children:"Runs end-to-end tests via make test-e2e"}),"\n",(0,s.jsx)(n.li,{children:"Passes if tests succeed"}),"\n",(0,s.jsx)(n.li,{children:"Fails if tests return a non-zero exit code"}),"\n"]}),"\n",(0,s.jsx)(n.h3,{id:"running-github-actions-locally",children:"Running GitHub Actions locally"}),"\n",(0,s.jsxs)(n.p,{children:["If we only run tests when creating a PR or pushing to ",(0,s.jsx)(n.code,{children:"main"}),", you may make mistakes and create many commits just to trigger testing. Here, we can use ",(0,s.jsx)(n.code,{children:"act"}),", a local command line tool that can help:"]}),"\n",(0,s.jsx)(n.admonition,{type:"info",children:(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["For more details on how to use ",(0,s.jsx)(n.code,{children:"act"}),": ",(0,s.jsx)(n.a,{href:"https://nektosact.com",children:"https://nektosact.com"})]}),"\n",(0,s.jsx)(n.li,{children:"Install Docker on your local machine first"}),"\n"]})}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:"# install\npixi add act\n# run e2e test locally, it will run this on your local machine\nact -W .github/workflows/test-e2e.yaml\n"})}),"\n",(0,s.jsx)(n.p,{children:(0,s.jsx)(n.img,{alt:"act_failed",src:i(7795).A+"",width:"1519",height:"864"})}),"\n",(0,s.jsx)(n.admonition,{type:"warning",children:(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["As you can see, it failed because pixi is not installed. This is because the ",(0,s.jsx)(n.code,{children:"make test-e2e"})," target does not trigger pixi installation"]}),"\n",(0,s.jsxs)(n.li,{children:["Also, there is a wrong URL for installing pixi: ",(0,s.jsx)(n.code,{children:"https://pixi.dev/install.sh"})," should be replaced with ",(0,s.jsx)(n.code,{children:"https://pixi.sh/install.sh"})]}),"\n",(0,s.jsxs)(n.li,{children:["You should add pixi installation as a dependency for ",(0,s.jsx)(n.code,{children:"test-e2e"})]}),"\n"]})}),"\n",(0,s.jsxs)(n.p,{children:["Update dependencies of ",(0,s.jsx)(n.code,{children:"test-e2e"})," in the ",(0,s.jsx)(n.code,{children:"Makefile"}),":"]}),"\n",(0,s.jsx)(n.pre,{children:(0,s.jsx)(n.code,{className:"language-bash",children:".PHONY: test-e2e clean\n${HOME}/.pixi/bin/pixi:\n\tmkdir -p ${HOME}/.pixi/bin\n\tcurl -sSL https://pixi.sh/install.sh | sh\n\ndata/GSE110004:\n\tmkdir -p data/GSE110004\n\tcd data/GSE110004 && \\\n\twget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_1.fastq.gz && \\\n\twget https://raw.githubusercontent.com/nf-core/test-datasets/rnaseq/testdata/GSE110004/SRR6357070_2.fastq.gz\n\ntest-e2e: data/GSE110004 ${HOME}/.pixi/bin/pixi\n\t${HOME}/.pixi/bin/pixi run bash main.sh data/GSE110004/*_1.fastq.gz data/GSE110004/*_2.fastq.gz result\n\nclean:\n\trm -rf data/GSE110004\n"})}),"\n",(0,s.jsxs)(n.p,{children:["Now you can rerun to test successfully, as shown below:\n",(0,s.jsx)(n.img,{alt:"act_success1",src:i(4246).A+"",width:"1519",height:"864"})]}),"\n",(0,s.jsx)(n.p,{children:(0,s.jsx)(n.img,{alt:"act_success2",src:i(6013).A+"",width:"1519",height:"864"})}),"\n",(0,s.jsx)(n.h2,{id:"what-is-cicd-in-bioinformatics",children:"What is CI/CD in bioinformatics?"}),"\n",(0,s.jsx)(n.h3,{id:"what-is-cicd-in-software-development",children:"What is CI/CD in software development"}),"\n",(0,s.jsxs)(n.p,{children:[(0,s.jsx)(n.img,{alt:"cicd",src:i(5623).A+"",width:"825",height:"399"}),"\nReference: ",(0,s.jsx)(n.a,{href:"https://viblo.asia/p/cau-hinh-cicd-voi-github-phan-1-mot-it-ly-thuyet-Qbq5Q9NL5D8",children:"https://viblo.asia/p/cau-hinh-cicd-voi-github-phan-1-mot-it-ly-thuyet-Qbq5Q9NL5D8"})]}),"\n",(0,s.jsx)(n.p,{children:"What exactly are CI (Continuous Integration) and CD (Continuous Delivery)? These terms are commonly used in software development.\nIn the figure above, software can be developed with many new features, which will trigger specific jobs with actions.\nFor example, we can the workflow to clone the repository on a GitHub Actions instance and run tests. This approach depends on many use cases."}),"\n",(0,s.jsx)(n.h3,{id:"how-does-it-work-in-bioinformatics",children:"How does it work in bioinformatics?"}),"\n",(0,s.jsx)(n.p,{children:"That is exactly what I showed you above. If you want to develop software, follow the standard CI/CD software development process. In contrast,\nbioinformatics typically combines these tools together. This means they will run scripts (bash, Nextflow, WDL, Snakemake, etc.) to combine tools\nin a workflow manner. The release is the repository itself with the scripts inside (CD), while the CI is the process we set up to run\nthe pipeline with testing datasets."}),"\n",(0,s.jsxs)(n.p,{children:["Now you can create a PR, which will always trigger the testing that helps you validate whether your changes may fail the pipeline.\n",(0,s.jsx)(n.img,{alt:"PR_GHA",src:i(798).A+"",width:"1061",height:"1071"})]}),"\n",(0,s.jsx)(n.admonition,{type:"tip",children:(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["Use secrets in ",(0,s.jsx)(n.code,{children:"Settings"})," to work with credentials (passwords, usernames, keys)"]}),"\n",(0,s.jsx)(n.li,{children:"Use the Makefile to help sync the local environment with the GitHub Actions environment. This makes it easy to change the CI/CD runner"}),"\n",(0,s.jsx)(n.li,{children:"Use matrix strategy to run tests faster"}),"\n"]})}),"\n",(0,s.jsxs)(n.p,{children:["Check GitHub Actions logs for debugging:\n",(0,s.jsx)(n.img,{alt:"GHA",src:i(1366).A+"",width:"1381",height:"1356"})]}),"\n",(0,s.jsx)(n.h2,{id:"review",children:"Review"}),"\n",(0,s.jsx)(n.p,{children:"In this tutorial, we explored how to transform a basic bioinformatics project into a reproducible, testable pipeline using Git, Makefiles, and CI/CD practices. Here are the key takeaways:"}),"\n",(0,s.jsx)(n.h3,{id:"git-commands-for-project-management",children:"Git Commands for Project Management"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["Use ",(0,s.jsx)(n.code,{children:"git log"})," and ",(0,s.jsx)(n.code,{children:"git log --oneline"})," to view commit history"]}),"\n",(0,s.jsx)(n.li,{children:"Checkout to specific commits to track development progress"}),"\n",(0,s.jsxs)(n.li,{children:["Use ",(0,s.jsx)(n.code,{children:"git diff"})," to compare changes between commits or branches"]}),"\n",(0,s.jsxs)(n.li,{children:["Roll back to original states with ",(0,s.jsx)(n.code,{children:"git checkout -- ."})]}),"\n",(0,s.jsxs)(n.li,{children:["Copy files from other branches using ",(0,s.jsx)(n.code,{children:"git checkout <branch> <file>"})]}),"\n"]}),"\n",(0,s.jsx)(n.h3,{id:"makefile-automation",children:"Makefile Automation"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["Create targets for data download (",(0,s.jsx)(n.code,{children:"make data"}),") and cleanup (",(0,s.jsx)(n.code,{children:"make clean"}),")"]}),"\n",(0,s.jsxs)(n.li,{children:["Add test targets (",(0,s.jsx)(n.code,{children:"make test-e2e"}),") that chain dependencies together"]}),"\n",(0,s.jsx)(n.li,{children:"Ensure reproducibility by standardizing commands across environments"}),"\n"]}),"\n",(0,s.jsx)(n.h3,{id:"workflow-scripts",children:"Workflow Scripts"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Create bash scripts that accept command-line arguments"}),"\n",(0,s.jsx)(n.li,{children:"Combine tools like FastQC and MultiQC in a workflow"}),"\n",(0,s.jsx)(n.li,{children:"Test workflows locally before pushing to remote"}),"\n"]}),"\n",(0,s.jsx)(n.h3,{id:"cicd-with-github-actions",children:"CI/CD with GitHub Actions"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["Create workflow files in ",(0,s.jsx)(n.code,{children:".github/workflows/"})]}),"\n",(0,s.jsx)(n.li,{children:"Trigger tests on push to main or pull requests"}),"\n",(0,s.jsx)(n.li,{children:"Use the same Makefile commands in CI as in local development"}),"\n",(0,s.jsx)(n.li,{children:"Check GitHub Actions logs for debugging failed tests"}),"\n"]}),"\n",(0,s.jsxs)(n.h3,{id:"local-testing-with-act",children:["Local Testing with ",(0,s.jsx)(n.code,{children:"act"})]}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:["Test GitHub Actions workflows locally using ",(0,s.jsx)(n.code,{children:"act"})]}),"\n",(0,s.jsx)(n.li,{children:"Catch issues before pushing to remote repositories"}),"\n",(0,s.jsx)(n.li,{children:"Reduce the number of commits needed to fix CI failures"}),"\n"]}),"\n",(0,s.jsx)(n.h3,{id:"cicd-in-bioinformatics-context",children:"CI/CD in Bioinformatics Context"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsx)(n.li,{children:"Unlike traditional software, bioinformatics CI/CD often focuses on pipeline validation"}),"\n",(0,s.jsx)(n.li,{children:'The "delivery" is often the repository itself with reproducible workflows'}),"\n",(0,s.jsx)(n.li,{children:"Testing involves running entire pipelines on test datasets"}),"\n",(0,s.jsx)(n.li,{children:"This ensures changes don't break existing analyses"}),"\n"]}),"\n",(0,s.jsx)(n.p,{children:"By implementing these practices, you create a robust bioinformatics project that is:"}),"\n",(0,s.jsxs)(n.ul,{children:["\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Reproducible"}),": Anyone can reproduce your results on any machine"]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Testable"}),": Automated tests catch errors before they reach production"]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Maintainable"}),": Clear history and documentation make updates easier"]}),"\n",(0,s.jsxs)(n.li,{children:[(0,s.jsx)(n.strong,{children:"Collaborative"}),": CI/CD ensures all changes are validated before merging"]}),"\n"]}),"\n",(0,s.jsx)(n.p,{children:"Ready to implement CI/CD in your own bioinformatics projects? Start by adding a simple Makefile target, then gradually build up your testing and automation capabilities."})]})}function h(e={}){const{wrapper:n}={...(0,a.R)(),...e.components};return n?(0,s.jsx)(n,{...e,children:(0,s.jsx)(d,{...e})}):d(e)}},1366(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/GHA_log-275ac35c8fdbaf250724faba33aaee7c.png"},1595(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/qc_workflow-9a92452d2aaa1a2b5a25d1da3c212a57.png"},4051(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/github_intro-d1dddb912f0928fbf8ff52fa2d7db7b4.svg"},4246(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/act_success1-c6c94a384f9f52c48c288f723338945c.png"},5623(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/cicd-85c944e1969870ad96df0a1b7d04d315.webp"},6013(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/act_success2-66c5e05f88d51a5c3604ba0fa4fd927a.png"},7795(e,n,i){i.d(n,{A:()=>t});const t=i.p+"assets/images/act_failed-afc9d4fa6bcd8fd64420bfe5598a61d5.png"},8453(e,n,i){i.d(n,{R:()=>o,x:()=>r});var t=i(6540);const s={},a=t.createContext(s);function o(e){const n=t.useContext(a);return t.useMemo(function(){return"function"==typeof e?e(n):{...n,...e}},[n,e])}function r(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(s):e.components||s:o(e.components),t.createElement(a.Provider,{value:n},e.children)}},9264(e){e.exports=JSON.parse('{"permalink":"/river-docs/blog/how-to-version-control-git-bioinformatics-part-2","source":"@site/blog/2026-01/2026-01-16.md","title":"The Evolution of Version Control - Git\'s Role in Reproducible Bioinformatics (Part 2)","description":"Welcome to Part 2 of our series on version control in bioinformatics. In Part 1, we introduced Git fundamentals, branching strategies, and collaborative workflows. In this post, we\'ll dive into how Continuous Integration and Continuous Deployment (CI/CD) can transform your bioinformatics projects. If these concepts are new to you, don\'t worry\u2014this guide will walk you through managing your bioinformatics repository to ensure your work is easily reproducible on any machine. Whether your server is wiped or you need to spin up a new virtual machine, you\'ll be able to quickly rerun your pipeline. With CI/CD, every code update can automatically trigger tests on a small dataset to verify everything works before scaling up, ensuring that new changes don\'t break your results or workflows.","date":"2026-01-16T00:00:00.000Z","tags":[{"inline":true,"label":"git","permalink":"/river-docs/blog/tags/git"},{"inline":true,"label":"version-control","permalink":"/river-docs/blog/tags/version-control"},{"inline":true,"label":"bioinformatics","permalink":"/river-docs/blog/tags/bioinformatics"},{"inline":true,"label":"reproducibility","permalink":"/river-docs/blog/tags/reproducibility"},{"inline":true,"label":"pipelines","permalink":"/river-docs/blog/tags/pipelines"},{"inline":true,"label":"ci-cd","permalink":"/river-docs/blog/tags/ci-cd"},{"inline":true,"label":"github-actions","permalink":"/river-docs/blog/tags/github-actions"}],"readingTime":13.21,"hasTruncateMarker":true,"authors":[{"name":"Thanh-Giang Tan Nguyen","title":"Founder at RIVER","url":"https://www.facebook.com/nttg8100","page":{"permalink":"/river-docs/blog/authors/river"},"email":"nttg8100@gmail.com","socials":{"linkedin":"https://www.linkedin.com/in/thanh-giang-tan-nguyen-761b28190/","github":"https://github.com/nttg8100"},"imageURL":"https://avatars.githubusercontent.com/u/64969412?v=4","key":"river"}],"frontMatter":{"slug":"how-to-version-control-git-bioinformatics-part-2","title":"The Evolution of Version Control - Git\'s Role in Reproducible Bioinformatics (Part 2)","authors":["river"],"tags":["git","version-control","bioinformatics","reproducibility","pipelines","ci-cd","github-actions"],"image":"./imgs/github_intro.svg"},"unlisted":false,"prevItem":{"title":"Bioinformatics Cost Optimization for Computing Resources Using Nextflow (Part 1)","permalink":"/river-docs/blog/bioinformatics-computing-resource-optimization-part1"},"nextItem":{"title":"The Evolution of Version Control - Git\'s Role in Reproducible Bioinformatics (Part 1)","permalink":"/river-docs/blog/how-to-version-control-git-bioinformatics-part-1"}}')}}]);