---
slug: machine-learning-bioinformatics-part2-knn
title: "Machine Learning in Bioinformatics Part 2: Using KNN for early detection of breast cancer"
authors: [river]
tags: [machine-learning, bioinformatics, python, knn, gene-expression]
image: ./imgs/intro.png
draft: true
---

<!-- truncate -->
## Dataset
### Summary
The data set is collected from this study, registered on NCBI GEO database with ID GSE211692: A large-scale study of 9,921 patients with 13 types of human solid cancers. 
Built with Streamlit, it enables interactive exploration, group comparisons, and generation of 
publication-ready visualizations. It is the microarray dataset with designed from the miRNA gene expression measurement.
 
:::info
Matsuzaki J, Kato K, Oono K, Tsuchiya N, Sudo K, Shimomura A, Tamura K, Shiino S, Kinoshita T, Daiko H, Wada T, Katai H, Ochiai H, Kanemitsu Y, Takamaru H, Abe S, Saito Y, Boku N, Kondo S, Ueno H, Okusaka T, Shimada K, Ohe Y, Asakura K, Yoshida Y, Watanabe SI, Asano N, Kawai A, Ohno M, Narita Y, Ishikawa M, Kato T, Fujimoto H, Niida S, Sakamoto H, Takizawa S, Akiba T, Okanohara D, Shiraishi K, Kohno T, Takeshita F, Nakagama H, Ota N, Ochiya T; Project Team for Development and Diagnostic Technology for Detection of miRNA in Body Fluids. Prediction of tissue-of-origin of early stage cancers using serum miRNomes. JNCI Cancer Spectr. 2023 Jan 3;7(1):pkac080. doi: 10.1093/jncics/pkac080. PMID: 36426871; PMCID: PMC9825310.`
:::

### Preprocess
The orginal paper provides the specific restriction setup. Here, I forked a new repository that helps:
+ Use pixi to setup the consistent environment. No need to manually set up if you already have pixi
+ Integrate with RIVER platform where you can run it on HPC and output the processed data to the platform project.
+ Automatically download raw data, aggregate the data and re-label, it can be ready for analysis or machine learning modeling
  
You may not have configure RIVER platform, run it locally is sufficient via

```bash
git clone https://github.com/riverxdata/river-head_model
# run e2e
cd river-head_model
bash tests/run_local.sh

# check your output, large files are put into the .gitignore
ls -lah  work/jobs/job_id/analysis/preprocess
# total 1,7G
# drwxrwxr-x  7 giangnguyen giangnguyen 1,3M Thg 2   7 13:24 .
# drwxrwxr-x 10 giangnguyen giangnguyen 4,0K Thg 2   7 13:00 ..
# -rw-rw-r--  1 giangnguyen giangnguyen 737M Thg 2   7 13:24 all_data.csv
# drwxrwxr-x  3 giangnguyen giangnguyen 4,0K Thg 2   7 10:41 app
# -rw-rw-r--  1 giangnguyen giangnguyen   51 Thg 2   7 13:24 category_distribution.csv
# -rwxrwxr-x  1 giangnguyen giangnguyen   82 Thg 2   7 13:25 .gitignore
# -rw-rw-r--  1 giangnguyen giangnguyen 981M Thg 2   7 11:07 GSE211692_RAW.tar
# -rw-rw-r--  1 giangnguyen giangnguyen  493 Thg 2   7 13:24 label_distribution.csv
# drwxrwxr-x  7 giangnguyen giangnguyen 4,0K Thg 2   7 10:41 lib
# -rw-rw-r--  1 giangnguyen giangnguyen 1,2M Thg 2   7 13:00 mapping_file.txt
# -rwxrwxr-x  1 giangnguyen giangnguyen 1,8K Thg 2   7 10:41 README.md
# -rwxrwxr-x  1 giangnguyen giangnguyen  153 Thg 2   7 10:41 run_dann.sh
# -rwxrwxr-x  1 giangnguyen giangnguyen 1,2K Thg 2   7 13:04 run_preprocess.sh
# -rwxrwxr-x  1 giangnguyen giangnguyen  498 Thg 2   7 10:41 setup.py
# drwxrwxr-x  6 giangnguyen giangnguyen 4,0K Thg 2   7 10:41 test
# drwxrwxr-x  2 giangnguyen giangnguyen 4,0K Thg 2   7 13:00 tool
# drwxrwxr-x  8 giangnguyen giangnguyen 4,0K Thg 2   7 13:21 work_dir
```

For this dataset, each row is a sample, each column is the feature. Beside, it has the additional column for the label of each sample that will be used for the machine learning classification

The basic summary is also ran on the above step, for detail, it will count the value of each group samples.

## Environment
For all of the tutorial from RIVERXDATA, to ensure reliabilty and reproducibility, I use pixi or container (docker, singularity).
```bash
pixi init
pixi workspace channel add conda-forge bioconda
pixi workspace platform add osx-arm64 linux-64
# add tools
pixi add python=3.12 jupyterlab
pixi add --pypi pandas
```



